cardano-node-emulator/src/Cardano/Node/Emulator/MTL.hs

```diff
@@ -0,0 +1,221 @@
+      Nothing
+    toDecoratedDatum (C.TxOutDatumHash _ h) =
+      Just (PV2.DatumHash $ PlutusTx.toBuiltin (C.serialiseToRawBytes h), DatumUnknown)
+    toDecoratedDatum (C.TxOutDatumInTx _ d) =
+      Just (PV2.DatumHash $ PlutusTx.toBuiltin (C.serialiseToRawBytes (C.hashScriptData d)), DatumInBody $ PV2.Datum $ fromCardanoScriptData d)
+    toDecoratedDatum (C.TxOutDatumInline _ d) =
+      Just (PV2.DatumHash $ PlutusTx.toBuiltin (C.serialiseToRawBytes (C.hashScriptData d)), DatumInline $ PV2.Datum $ fromCardanoScriptData d)
+
+-- | Query the total value of the unspent transaction outputs at the given address.
+fundsAt :: MonadEmulator m => CardanoAddress -> m C.Value
+fundsAt addr = foldMap (view decoratedTxOutValue) <$> utxosAt addr
+
+
+-- | Balance an unbalanced transaction, using funds from the given wallet if needed, and returning any remaining value to the same wallet.
+balanceTx
+  :: MonadEmulator m
+  => UtxoIndex -- ^ Just the transaction inputs, not the entire 'UTxO'.
+  -> CardanoAddress -- ^ Wallet address
+  -> CardanoBuildTx
+  -> m CardanoTx
+balanceTx utxoIndex changeAddr utx = do
+  params <- ask
+  es <- get
+  let
+    ownUtxos = UtxoIndex $ snd <$> es ^. esAddressMap . AM.fundsAt changeAddr
+    utxoProvider = E.utxoProviderFromWalletOutputs ownUtxos utx
+  CardanoEmulatorEraTx <$> E.makeAutoBalancedTransactionWithUtxoProvider
+      params
+      utxoIndex
+      changeAddr
+      (either (throwError . BalancingError) pure . utxoProvider)
+      (throwError . either ValidationError ToCardanoError)
+      utx
+
+-- | Balance a transaction, sign it with the given signatures, and finally queue it.
+submitUnbalancedTx
+  :: (MonadEmulator m, Foldable f)
+  => UtxoIndex -- ^ Just the transaction inputs, not the entire 'UTxO'.
+  -> CardanoAddress -- ^ Wallet address
+  -> CardanoBuildTx
+  -> f PaymentPrivateKey -- ^ Signatures
+  -> m CardanoTx
+submitUnbalancedTx utxoIndex changeAddr utx keys = do
+  newTx <- balanceTx utxoIndex changeAddr utx
+  let signedTx = foldr (addCardanoTxSignature . unPaymentPrivateKey) newTx keys
+  queueTx signedTx
+  pure signedTx
+
+-- | Create a transaction that transfers funds from one address to another, and sign and submit it.
+payToAddress :: MonadEmulator m => (CardanoAddress, PaymentPrivateKey) -> CardanoAddress -> C.Value -> m PV2.TxId
+payToAddress (sourceAddr, sourcePrivKey) targetAddr value = do
+  let buildTx = CardanoBuildTx $ E.emptyTxBodyContent
+           { C.txOuts = [C.TxOut targetAddr (toCardanoTxOutValue value) C.TxOutDatumNone C.ReferenceScriptNone]
+           }
+  getCardanoTxId <$> submitUnbalancedTx mempty sourceAddr buildTx [sourcePrivKey]
+
+
+-- | Test the number of validated transactions and the total number of transactions.
+-- Returns a failure message if the numbers don't match up.
+hasValidatedTransactionCountOfTotal :: Int -> Int -> Seq E.ChainEvent -> Maybe String
+hasValidatedTransactionCountOfTotal valid total lg =
```

Comment:
Isn't it for test purpose only? If yes, did you consider moving it into a test module?